Mathlib/Tactic/Isolate/Core.lean

```diff
@@ -0,0 +1,345 @@
+  let xArgArray := args.filter fun arg ↦ arg.hasLooseBVar 0
+  match xArgArray with
+  | #[xArg] =>
+    let idx := args.findIdx (· == xArg)
+    -- Look up the `@[isolate]` lemmas with the right relation, function, argument index and
+    -- LHS/RHS positioning.
+    let key : IsolateLemmaKey := ⟨relName, fnName, idx, !symmetric? && !lhsContains⟩
+    let isolateDict := isolateExt.getState (← getEnv)
+    let lemmas := isolateDict.getD key #[]
+    let s ← saveState
+    -- Loop over the lemmas to see if any apply:
+    for lem in lemmas do
+      try
+        trace[Meta.isolate] "trying the lemma {lem} to rewrite the expression {P'}"
+        -- An `@[isolate]` lemma has the structure `_ → _ → ... → _ → (lemLHS ↔ lemRHS)`; parse this
+        let e ← mkConstWithFreshMVarLevels lem
+        let eTy ← inferType e
+        let (args, _, ty) ← forallMetaTelescopeReducing eTy
+        let .app (.app _ lemLHS) lemRHS := ty | throwError "ill-formed @[isolate] lemma {lem}"
+        -- Attempt to unify `lemLHS` with the expression `P'` being worked on.
+        guard (← isDefEq P' lemLHS)
+        -- If that succeeded, we know what `P'` will be transformed to, namely the instantiated RHS
+        -- of the lemma.
+        let Q ← instantiateMVars lemRHS
+        -- Collect the unassigned metavariables, i.e. side goals.
+        let mvars := (← args.mapM getMVars).flatten
+        -- We attempt to resolve each of these side goals;
+        -- we collect those side goals which are not resolved.
+        let mut unresolvedMVars : Array MVarId := #[]
+        for mvar in mvars do
+          try
+            -- attempt to solve the side goal by typeclass inference
+            mvar.inferInstance
+          catch _ => try
+            -- attempt to solve the side goal using `gcongr`'s discharger (currently this is a
+            -- wrapper for `positivity`)
+            GCongr.gcongrDischarger mvar
+          catch _ =>
+            unresolvedMVars := unresolvedMVars.push mvar
+        -- Make the proof of `P' ↔ Q` (dependent on these side goals) and send it back.
+        let pf ← mkAppM ``propext #[← mkAppOptM lem (args.map some)]
+        return (unresolvedMVars.toList, ← symmResult.mkEqTrans { expr := Q, proof? := some pf })
+      catch _ => s.restore
+    throwError "no suitable lemmas found"
+  | _ => throwError "{x} is not localized to a single argument of {e}"
+
+/-- Recursively modify the expression `P`, which must be of the form `A ~ B`, by finding the
+expression `x` as a subexpression of either `A` or `B` (but not both), and rewriting by `@[isolate]`
+lemmas to make `x` less deeply nested.
+
+Return the list of side goals generated by the `@[isolate]` lemmas used (and which are not
+discharged by unification, type class inference, or the standard discharger), together with a
+`Simp.Result` (containing the new expression and a proof of its logical equivalence to `P`). -/
```

Comment:
Let's explain when this function results in a failure state.
```suggestion
`Simp.Result` (containing the new expression and a proof of its logical equivalence to `P`).

Fails if no steps could be made, otherwise succeeds even `x` is not fully isolated.
-/
```